rust/sbd-client/src/lib.rs

```diff
@@ -0,0 +1,320 @@
+//! Sbd client library.
+#![deny(missing_docs)]
+
+use std::io::{Error, Result};
+use std::sync::Arc;
+
+/// defined by the sbd spec
+const MAX_MSG_SIZE: usize = 20_000;
+
+#[cfg(feature = "raw_client")]
+pub mod raw_client;
+#[cfg(not(feature = "raw_client"))]
+mod raw_client;
+
+mod send_buf;
+
+/// Crypto to use. Note, the pair should be fresh for each new connection.
+pub trait Crypto {
+    /// The pubkey.
+    fn pub_key(&self) -> &[u8; 32];
+
+    /// Sign the nonce.
+    fn sign(&self, nonce: &[u8]) -> [u8; 64];
+}
+
+#[cfg(feature = "crypto")]
+mod default_crypto {
+    /// Default signer. Use a fresh one for every new connection.
+    pub struct DefaultCrypto([u8; 32], ed25519_dalek::SigningKey);
+
+    impl Default for DefaultCrypto {
+        fn default() -> Self {
+            let k =
+                ed25519_dalek::SigningKey::generate(&mut rand::thread_rng());
+            let pk = k.verifying_key().to_bytes();
+            Self(pk, k)
+        }
+    }
+
+    impl super::Crypto for DefaultCrypto {
+        fn pub_key(&self) -> &[u8; 32] {
+            &self.0
+        }
+
+        fn sign(&self, nonce: &[u8]) -> [u8; 64] {
+            use ed25519_dalek::Signer;
+            self.1.sign(nonce).to_bytes()
+        }
+    }
+}
+#[cfg(feature = "crypto")]
+pub use default_crypto::*;
+
+/// Public key.
+#[derive(Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct PubKey(pub [u8; 32]);
+
+impl std::fmt::Debug for PubKey {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        use base64::Engine;
+        let out = base64::engine::general_purpose::URL_SAFE_NO_PAD
+            .encode(&self.0[..]);
+        f.write_str(&out)
+    }
+}
+
+const CMD_FLAG: &[u8; 28] = &[0; 28];
+
+enum MsgType<'t> {
```

Comment:
could we encode a protocol version exchange message here or somewhere else?

Reply:
No protocol version. The protocol is designed to be extensible without versioning.